src/js/components/FocusedContainer.js

```diff
@@ -12,13 +12,81 @@ export const FocusedContainer = ({
 }) => {
   const [bodyOverflowStyle, setBodyOverflowStyle] = useState('');
   const ref = useRef(null);
+  const preNodeRef = useRef(null);
+  const postNodeRef = useRef(null);
 
-  const roots = useContext(RootsContext);
-  const [nextRoots, setNextRoots] = useState(roots);
+  const { roots: rootsContextValue } = useContext(RootsContext);
   useEffect(() => {
-    // make sure value of null is not added to array
-    if (ref.current) setNextRoots([...roots, ref.current]);
-  }, [roots]);
+    const container = ref.current;
+    const roots = rootsContextValue.current;
+
+    const handleTrapFocus = (e) => {
+      if (!hidden && trapFocus && container) {
+        const focusableElements = container.querySelectorAll(
+          `button:not([tabindex="-1"], [href]:not([tabindex="-1"], 
+     input:not([tabindex="-1"], select:not([tabindex="-1"], 
+     textarea:not([tabindex="-1"], [tabindex]:not([tabindex="-1"])`,
+        );
+
+        if (focusableElements.length === 0) return;
+
+        const firstElement = focusableElements[0];
+        const lastElement = focusableElements[focusableElements.length - 1];
+
+        // only perform focus if this is the most recently opened drop
+        if (roots[roots.length - 1] === container) {
+          if (container.contains(e.target)) container.lastFocus = e.target;
+          else if (container.lastFocus === firstElement) {
+            lastElement.focus();
+            e.preventDefault();
+          } else {
+            // In the case where the focus hasn't already been placed on
+            // or within the container, this will ensure the next "tab"
+            // places focus on the first focusable element
+            firstElement.focus();
+            e.preventDefault();
+          }
+        }
+      }
+    };
+
+    const addListeners = () => {
+      document.addEventListener('focus', handleTrapFocus, true);
+    };
+
+    const removeListeners = () => {
+      document.removeEventListener('focus', handleTrapFocus, true);
+    };
+
+    // add container to the global roots
+    if (container) roots.push(container);
+
+    // Create and insert focusable nodes to help track when focus
+    // has left this container but without letting focus be noticeably placed
+    // on anything outside the container
+    const preDiv = document.createElement('div');
+    const postDiv = document.createElement('div');
+
+    preNodeRef.current = container.parentNode.insertBefore(preDiv, container);
+    postNodeRef.current = container.parentNode.insertBefore(
+      postDiv,
+      container.nextSibling,
+    );
+    preNodeRef.current.tabIndex = 0;
+    postNodeRef.current.tabIndex = 0;
+
+    addListeners();
+
+    return () => {
+      // remove from global roots array
+      if (roots.indexOf(container)) roots.splice(roots.indexOf(container), 1);
```

Comment:
Try to simplify this down to only the roots part to start, to see if it resolves the issue.

Maybe separate out the useEffects with the preNode/postNode or check if they still exist before calling `.remove()`

Reply:
Also could explore using `.filter()` on the array instead of `splice`

Reply:
If we end up keeping the handler function here, then we might be able to initialize the RootsContext array to include `document` (to handle cases where Drop/Layer is rendered outside of Grommet tag).